internal/webhooks/cluster.go

```diff
@@ -395,6 +436,183 @@ func (webhook *Cluster) validateTopology(ctx context.Context, oldCluster, newClu
+			)
+		}
+		if mdVersion.NE(oldVersion) {
+			return field.Invalid(
+				fldPath,
+				*m.Spec.Template.Spec.Version,
+				fmt.Sprintf("blocking update because the version %q of MachineDeployment %s does not match the current version %q", mdVersion, m.GetName(), oldVersion),
+			)
+		}
+
+		mdTopologyName, ok := m.ObjectMeta.Labels[clusterv1.ClusterTopologyMachineDeploymentNameLabel]
+		if !ok || mdTopologyName == "" {
+			return fmt.Errorf("failed to find label %s in %s", clusterv1.ClusterTopologyMachineDeploymentNameLabel, tlog.KObj{Obj: m})
+		}
+
+		// Make sure that the name of the MachineDeployment stays unique.
+		// If we've already seen a MachineDeployment with the same name
+		// this is an error, probably caused from manual modifications or a race condition.
+		if _, ok := state[mdTopologyName]; ok {
+			return fmt.Errorf("duplicate %s found for label %s: %s", tlog.KObj{Obj: m}, clusterv1.ClusterTopologyMachineDeploymentNameLabel, mdTopologyName)
+		}
+
+		state[mdTopologyName] = &scope.MachineDeploymentState{
+			Object: m,
+		}
+	}
+
+	mdUpgradingNames, err := state.Upgrading(ctx, ctrlClient)
+	if err != nil {
+		return err
+	}
+
+	if len(mdUpgradingNames) > 0 {
+		return fmt.Errorf("upgrading MachineDeployments: [%s]", strings.Join(mdUpgradingNames, ", "))
+	}
+
+	return nil
+}
+
+func validateTopologyMachinePoolVersions(ctx context.Context, fldPath *field.Path, ctrlClient client.Reader, wlClient client.Reader, oldCluster *clusterv1.Cluster, oldVersion semver.Version) error {
+	// FROM: func (r *Reconciler) getCurrentMachineDeploymentState(ctx context.Context, blueprintMachineDeployments map[string]*scope.MachineDeploymentBlueprint, cluster *clusterv1.Cluster) (map[string]*scope.MachineDeploymentState, error) {
+	state := make(scope.MachinePoolsStateMap)
+
+	// List all the machine deployments in the current cluster and in a managed topology.
+	mp := &expv1.MachinePoolList{}
+	err := ctrlClient.List(ctx, mp,
+		client.MatchingLabels{
+			clusterv1.ClusterNameLabel:          oldCluster.Name,
+			clusterv1.ClusterTopologyOwnedLabel: "",
+		},
+		client.InNamespace(oldCluster.Namespace),
+	)
+	if err != nil {
+		return errors.Wrap(err, "failed to read MachinePools for managed topology")
+	}
+
+	for i := range mp.Items {
+		m := &mp.Items[i]
+
+		mpVersion, err := semver.ParseTolerant(*m.Spec.Template.Spec.Version)
+		if err != nil {
+			// NOTE: this should never happen. Nevertheless, handling this for extra caution.
+			return field.Invalid(
+				fldPath,
+				*m.Spec.Template.Spec.Version,
+				fmt.Sprintf("unable to parse version of MachinePool %s", m.GetName()),
+			)
+		}
+		if mpVersion.NE(oldVersion) {
+			return field.Invalid(
+				fldPath,
+				*m.Spec.Template.Spec.Version,
+				fmt.Sprintf("blocking update because the version %q of MachinePool %s does not match the current version %q", mpVersion, m.GetName(), oldVersion),
+			)
+		}
+
+		// TODO: I think we should not check this, its already in the topology controller.
```

Comment:
Let's talk about, but tl;dr. I think we don't care about the label so we can skip this stuff and also building the entire state